src/armature/Generator.ts

```diff
@@ -1,38 +1,189 @@
 
 type Definition = {
     weight: number;
-    generator(root: Point);
+    generator: GeneratorFn;
 };
 
 type SpawnPoint = {
     component: string;
     at: Point;
 };
 
+/**
+ * A cost function returns the cost of an entire model. The new nodes that were added
+ * since the last iteration of generation are passed to the cost function so that, if you can,
+ * you can just compute the incremental cost and add it to the previous cost, `instance.getCost()`.
+ */
+export type CostFn = (instance: GeneratorInstance, nodes: Node[]) => number;
+
+/**
+ * An instance of a generated model, possibly in the middle of generation.
+ */
+export class GeneratorInstance {
+    private model: Model = new Model();
+    private generator: Generator;
+    private costFn: CostFn;
+    private cost: number = 0;
+    private spawnPoints: SpawnPoint[] = [];
+    private random: RandomGenerator = Math.random;
+
+    /**
+     * Creates an instance of a generator, with a function specifying how to evaluate how "good"
+     * it is.
+     *
+     * @param {Generator} generator The generator this is an instance of.
+     * @param {CostFn} costFn A function to evaluate this instance.
+     */
+    constructor(generator: Generator, costFn: CostFn) {
+        this.generator = generator;
+        this.costFn = costFn;
+    }
+
+    /**
+     * A wrapper for `mode.add`, for convenience when defining rules.
+     *
+     * @param {Node} node The node to add to the model.
+     * @returns {Node} The node that was added.
+     */
+    public add(node: Node): Node {
+        return this.model.add(node);
+    }
+
+    /**
+     * @returns {GeneratorInstance} A flat copy of this generator instance.
+     */
+    public clone(): GeneratorInstance {
+        const cloned = new GeneratorInstance(this.generator, this.costFn);
+        cloned.model = this.model.clone();
+        cloned.cost = this.cost;
+        cloned.spawnPoints = [...this.spawnPoints];
+
+        return cloned;
+    }
+
+    /**
+     * @returns {Model} The model this instance has generated so far.
+     */
+    public getModel(): Model {
+        return this.model;
+    }
+
+    /**
+     * @returns {number} The current cost of the model the instance has generated so far.
+     */
+    public getCost(): number {
+        return this.cost;
+    }
+
+    /**
+     * @returns {number} How many spawn points the generator can choose from when growing the
+     * next generation.
+     */
+    public activeSpawnPoints(): number {
+        return this.spawnPoints.length;
+    }
+
+    /**
+     * Tells the generator that more components can be generated somewhere.
+     *
+     * @param {SpawnPoint} spawnPoint The name of the component to spawn and the point at which to
+     * spawn it.
+     */
+    public addDetail(spawnPoint: SpawnPoint) {
+        this.spawnPoints.push(spawnPoint);
+    }
+
+    /**
+     * Grows this instance, if possible, until a new shape is added.
+     */
+    public advance() {
```

Comment:
Hmm, `advance` might not be the most descriptive as to what's going on. Could we think of a different name, like `growInstance`, or `addShapeIfPossible`?

Reply:
maybe `growIfPossible`?

Reply:
I like that name 💯